File: src/main/java/org/team401/taxis/diffdrive/control/NonlinearFeedbackPathController.java
package org.team401.taxis.diffdrive.control;

import org.team401.taxis.geometry.Pose2d;
import org.team401.taxis.physics.DifferentialDrivetrainDynamics;
import org.team401.taxis.util.Units;
import org.team401.taxis.util.Util;

/**
 * @author Cameron Earle
 * @version 12/3/2018
 */
public class NonlinearFeedbackPathController implements PathController {
    public NonlinearFeedbackPathController(double kBeta, double kZeta) {
        this.kBeta = kBeta;
        this.kZeta = kZeta;
    }

    public NonlinearFeedbackPathController() {
        this(2.0, .7);
    }

    private double kBeta;
    private double kZeta;

    DifferentialDrivetrainDynamics.ChassisState prev_velocity_ = new DifferentialDrivetrainDynamics.ChassisState();

    @Override
    public DrivetrainPathManager.Output update(DifferentialDrivetrainDynamics.DriveDynamics dynamics, Pose2d current_state, Pose2d mError, DifferentialDrivetrainDynamics mModel, double mDt) {
        // Compute gain parameter.
        final double k = 2.0 * kZeta * Math.sqrt(kBeta * dynamics.chassis_velocity.linear * dynamics.chassis_velocity
                .linear + dynamics.chassis_velocity.angular * dynamics.chassis_velocity.angular);

        // Compute error components.
        final double angle_error_rads = mError.getRotation().getRadians();
        final double sin_x_over_x = Util.epsilonEquals(angle_error_rads, 0.0, 1E-2) ?
                1.0 : mError.getRotation().sin() / angle_error_rads;
        final DifferentialDrivetrainDynamics.ChassisState adjusted_velocity = new DifferentialDrivetrainDynamics.ChassisState(
                dynamics.chassis_velocity.linear * mError.getRotation().cos() +
                        k * Units.inches_to_meters(mError.getTranslation().x()),
                dynamics.chassis_velocity.angular + k * angle_error_rads +
                        dynamics.chassis_velocity.linear * kBeta * sin_x_over_x * Units.inches_to_meters(mError
                                .getTranslation().y()));

        // Compute adjusted left and right wheel velocities.
        dynamics.chassis_velocity = adjusted_velocity;
        dynamics.wheel_velocity = mModel.solveInverseKinematics(adjusted_velocity);

        dynamics.chassis_acceleration.linear = mDt == 0 ? 0.0 : (dynamics.chassis_velocity.linear - prev_velocity_
                .linear) / mDt;
        dynamics.chassis_acceleration.angular = mDt == 0 ? 0.0 : (dynamics.chassis_velocity.angular - prev_velocity_
                .angular) / mDt;

        prev_velocity_ = dynamics.chassis_velocity;

        DifferentialDrivetrainDynamics.WheelState feedforward_voltages = mModel.solveInverseDynamics(dynamics.chassis_velocity,
                dynamics.chassis_acceleration).voltage;

        return new DrivetrainPathManager.Output(dynamics.wheel_velocity.left, dynamics.wheel_velocity.right, dynamics.wheel_acceleration
                .left, dynamics.wheel_acceleration.right, feedforward_voltages.left, feedforward_voltages.right);
    }
}

File: src/main/kotlin/org/team401/taxis/diffdrive/NonlinearFeedbackPathController.kt
package org.team401.taxis.diffdrive

import org.team401.taxis.geometry.Pose2d
import org.team401.taxis.geometry.Pose2dWithCurvature
import org.team401.taxis.physics.DifferentialDrive
import org.team401.taxis.trajectory.TrajectoryIterator
import org.team401.taxis.trajectory.timing.TimedState
import org.team401.taxis.util.Units
import org.team401.taxis.util.Util

/**
 * @author Cameron Earle
 * @version 9/10/2018
 *
 */
class NonlinearFeedbackPathController(val kBeta: Double, val kZeta: Double): PathController {
    var prevVelocity = DifferentialDrive.ChassisState()

    override fun update(dt: Double, dynamics: DifferentialDrive.DriveDynamics, currentState: Pose2d, model: DifferentialDrive, trajectory: TrajectoryIterator<TimedState<Pose2dWithCurvature>>, setpoint: TimedState<Pose2dWithCurvature>, error: Pose2d, reversed: Boolean): Output {
        val k = 2.0 * kZeta * Math.sqrt(kBeta * dynamics.chassis_velocity.linear * dynamics.chassis_velocity
                .linear + dynamics.chassis_velocity.angular * dynamics.chassis_velocity.angular)

        val angleErrorRads = error.rotation.radians
        val sinXOverX = if (Util.epsilonEquals(angleErrorRads, 0.0, 1E-2)) 1.0
                        else error.rotation.sin() / angleErrorRads

        val adjustedVelocity =  DifferentialDrive.ChassisState(
                dynamics.chassis_velocity.linear * error.rotation.cos() +
                        k * Units.inches_to_meters(error.translation.x()),
                dynamics.chassis_velocity.angular + k * angleErrorRads +
                        dynamics.chassis_velocity.linear * kBeta * sinXOverX * Units.inches_to_meters(error
                        .translation.y()))

        dynamics.chassis_velocity = adjustedVelocity
        dynamics.wheel_velocity = model.solveInverseKinematics(adjustedVelocity)

        dynamics.chassis_acceleration.linear = if (dt == 0.0) 0.0
                                               else (dynamics.chassis_velocity.linear - prevVelocity.linear) / dt
        dynamics.chassis_acceleration.angular = if (dt == 0.0) 0.0
                                                else (dynamics.chassis_velocity.angular - prevVelocity.angular) / dt

        prevVelocity = dynamics.chassis_velocity

        val feedforwardVoltages = model.solveInverseDynamics(
                dynamics.chassis_velocity,
                dynamics.chassis_acceleration
        ).voltage

        return Output(dynamics.wheel_velocity.left, dynamics.wheel_velocity.right, dynamics.wheel_acceleration
                .left, dynamics.wheel_acceleration.right, feedforwardVoltages.left, feedforwardVoltages.right)
    }

    override fun reset() {
        prevVelocity = DifferentialDrive.ChassisState()
    }
}